### PR TITLE
[User management] Add race conditions test-tool tests for CreateUser and GrantUserRights [DPP-847]

### PR DESCRIPTION
**Pull request by pbatko-da**

### CODE REVIEW

File: ledger/error/src/main/scala/com/daml/error/ErrorCode.scala
Comment (mziolekda) [inline ledger/error/src/main/scala/com/daml/error/ErrorCode.scala:62]: This doesn't seem applicable in all circumstances

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala
Comment (mziolekda) [inline ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala:208]: When do we get an internal error?

Comment (pbatko-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala:208]: On db unique constraint violation

File: ledger/error/src/main/scala/com/daml/error/ErrorCode.scala
Comment (pbatko-da) [reply ledger/error/src/main/scala/com/daml/error/ErrorCode.scala:62]: Fixing

File: ledger/error/src/main/scala/com/daml/error/ErrorCode.scala
Comment (pbatko-da) [reply ledger/error/src/main/scala/com/daml/error/ErrorCode.scala:56]: Yes, for INTERNALs we deliberately strip all the information that would enable us to track down the corresponding error code

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala
Comment (tudor-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala:208]: Can we make this error deterministic?

File: ledger/error/src/main/scala/com/daml/error/ErrorCode.scala
Comment (tudor-da) [reply ledger/error/src/main/scala/com/daml/error/ErrorCode.scala:56]: This might be confusing to users of this API if they're not intimate with the way error codes work. Also, this function would better fit in a client-side library since it should be only used in testing or in client applications.

Comment (pbatko-da) [reply ledger/error/src/main/scala/com/daml/error/ErrorCode.scala:56]: I could move it somewhere into `//ledger/error/src/test/lib` but then it would have to be a harder to find and less pretty two argument function.

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala
Comment (tudor-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala:208]: I am not aware of the low-level details of the DB queries for this use-case, but wouldn't something like `INSERT.. ON CONFLICT DO NOTHING` and checking the returned value give us the basis for constructing the expected error on duplicate?

File: ledger/error/src/main/scala/com/daml/error/ErrorCode.scala
Comment (tudor-da) [reply ledger/error/src/main/scala/com/daml/error/ErrorCode.scala:56]: Related to the location of this utility function, the core error codes API (in `//ledger/error`) has two functions (intertwined unfortunately, but that's another story): 
1. Define the interfaces of error code DTOs
2. Constructs gRPC-domain objects to be sent to the client
The function added in this PR is about destructuring a gRPC-domain object which is the opposite of 2, so I would see it not well-placed here.

I would prefer this integrated in some other target/utility object at least that is aimed for clients. However, I wouldn't block this PR on this if you consider otherwise. 

In any case, please update the comment and/or method signature to highlight its restriction to non-security-sensitive errors.

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala
Comment (pbatko-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala:208]: `INSERT.. ON CONFLICT DO NOTHING` worked with Postgres but it wasn't obvious in Oracle
@nmarton-da have we figured out to replicate `INSERT.. ON CONFLICT DO NOTHING` in Oracle?
No idea about H2.

It's enticing to not assert receiving an internal error but we would likely have to pay a cost of diverging backend implementations.

File: ledger/error/src/main/scala/com/daml/error/ErrorCode.scala
Comment (pbatko-da) [reply ledger/error/src/main/scala/com/daml/error/ErrorCode.scala:60]: Added comment explaining the limitations

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala
Comment (tudor-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala:208]: Discussed offline. I don't think it's sane to return `INTERNAL` to the user easily on admin requests, but I'm fine to let it go if it's for a performance benefit/code simplicity. CC @nmarton-da @mziolekda

**Comment by tudor-da** [comment]
Had some concern regarding presenting a common error message to the admin as `INTERNAL` but as I understand it was agreed to do the implementation as is. Since this code is just asserting the implementation, I'll let it go and defer to @mziolekda or @nmarton-da

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala
Comment (nmarton-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala:209]: **TLDR**: with Pawel we discussed this back in the day, and this tradeoff was struck to 
* favor the majority of the use cases, 
* over a fragile/complex heuristic favoring a minor usage as well.

The fundamental idea behind this approach was discussed at the inception of the UserManagementStore:
* Optimistic query and set mechanics providing nice error messages without DB race (majority of cases)
* Providing general retry-able error (I guess that is the internal error case here) for DB races (expected to be vast minority of the cases)
* Please note there is no compromise on correctness here. It is only about nice error messages.

This tradeoff allowed not caring about special mechanics how we can achieve nice error messages in DB race cases.
Additional note: ON CONFLICT DO NOTHING works only nice for PostgreSQL, on Oracle it is a buggy bleeding edge, and on H2 for sure it will boil down to exception parsing to our best knowledge.

Comment (tudor-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala:209]: Alright, since you're saying this cannot be easily triggered in prod use-cases, I'm fine with it like this.

Comment (pbatko-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala:209]: @nmarton-da 

> Providing general retry-able error (I guess that is the internal error case here) for DB races (expected to be vast minority of the cases)

It's not retry-able. It is an opaque internal error without any details. The error code (visible only on the server side) is `INDEX_DB_SQL_NON_TRANSIENT_ERROR`


Fun fuct, on H2 contraint violation actually gets mapped to a retryable `INDEX_DB_SQL_TRANSIENT_ERROR`

Comment (tudor-da) [reply ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/UserManagementServiceIT.scala:209]: Yep, on H2 there's a caveat: https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/DatabaseSelfServiceError.scala#L32-L36

**Review by mziolekda — APPROVED** [review]

**Pull request merged**
